File: 백준/Gold/2294. 동전 2/동전 2.py
n,k = map(int, input().split()) # n : 동전 종류 k : 만들어야 하는 금액
coins = [int(input()) for _ in range(n)]
coins.sort()
dp = [[float('inf')] * (k+1) for _ in range(n+1)]
# dp[i][j] : i번째까지의 동전을 사용해 j원을 만들 때 사용된 동전의 최소 개수
for i in range(1 , n+1):
    for j in range(1, k+1):
        if j%coins[i-1] == 0 :
            dp[i][j] =  j//coins[i-1]
        elif j-coins[i-1] > 0:
            dp[i][j] = min(dp[i-1][j],dp[i][j], dp[i][j-coins[i-1]] + 1)
        else:
            dp[i][j] = dp[i-1][j]


print(dp[n][k] if dp[n][k] != float('inf') else -1)

File: 백준/Gold/2294. 동전 2/동전 2.java
import java.util.*;
import java.io.*;
public class Main {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st = new StringTokenizer(br.readLine());
    int n = Integer.parseInt(st.nextToken());
    int k = Integer.parseInt(st.nextToken());
    Set<Integer> coins = new HashSet<>();
    for(int i = 0 ; i < n ; i ++){
      coins.add(Integer.parseInt(br.readLine()));
    }
    int[] coinArr = new int[coins.size()];
    int idx = 0;
    for(int coin : coins){
      coinArr[idx++] = coin;
    }
    Arrays.sort(coinArr);
    int[][] dp = new int[k+1][coinArr.length+1];
    for(int i = 1 ; i <= k ; i ++){
      Arrays.fill(dp[i], 10001);
    }

    for(int i = 1 ; i <= k ; i ++){
      for(int j = 1 ; j <= coinArr.length ; j ++){
        int coin = coinArr[j-1];
        if(i < coin){
          dp[i][j] = dp[i][j-1];
          continue;
        }
        dp[i][j] = Math.min(dp[i][j-1], dp[i-coin][j] + 1);

      }
    }
    if (dp[k][coinArr.length] != 10001)
      System.out.println(dp[k][coinArr.length]);
    else
      System.out.println(-1);
  }

}